### PR TITLE
[Improvement] Support to config byte sizes by readable string (e.g., kb/mb/gb)

### PR DESCRIPTION
**Pull request by frankliee**
### What changes were proposed in this pull request?

Support to config byte sizes by readable string (e.g., kb/mb/gb).

### Why are the changes needed?

Use number (e.g., 100000) is unreadable and easy to make mistakes.

### Does this PR introduce _any_ user-facing change?

Yes.
In Java code: shuffleServerConf.getSizeAsBytes(ShuffleServerConf.SERVER_BUFFER_CAPACITY)
In conf file: rss.server.buffer.capacity 12kb

### How was this patch tested?

see com.tencent.rss.server.ShuffleServerConfTest

### CODE REVIEW

File: common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
```
Comment (jerqi) [inline common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java:17]: should we have blank line here?

Comment (frankliee) [reply common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java:17]: Fixed

File: server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java [hunk 2]
```diff
@@ -42,9 +42,9 @@
       .noDefaultValue()
       .withDescription("Spill threshold for buffer manager, it must be less than rss.server.buffer.capacity");
 
-  public static final ConfigOption<Integer> SERVER_PARTITION_BUFFER_SIZE = ConfigOptions
+  public static final ConfigOption<Long> SERVER_PARTITION_BUFFER_SIZE = ConfigOptions
       .key("rss.server.partition.buffer.size")
-      .intType()
+      .longType()
```
Comment (jerqi) [inline server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java:47]: like SERVER_READ_BUFFER_CAPACITY, many configs's default value is long. Should we replace these with kb.mb.gb.

Comment (frankliee) [reply server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java:47]: This patch only addresses size in bytes, in which long is required.

File: common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java [hunk 3]
```diff
@@ -95,7 +97,20 @@ static Long convertToLong(Object o) {
     } else if (o.getClass() == Integer.class) {
       return ((Integer) o).longValue();
     }
-    return Long.parseLong(o.toString());
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java:98]: convertToInt also need UnitConverter.

Comment (frankliee) [reply common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java:98]: Int type is used for other purposes (e.g., core numbers, replicas), rather size.

File: server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java (hunk 2 above)
Comment (frankliee) [reply server/src/main/java/com/tencent/rss/server/ShuffleServerConf.java:47]: Fixed in config file

**Comment by colinmjj** [comment]
@frankliee please resolve the conversations

File: common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java (hunk 3 above)
Comment (jerqi) [reply common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java:98]: int type may also be used for size.

Comment (frankliee) [reply common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java:98]: Using int type for size is unsafe.
So, we should use long type for all size types

Comment (jerqi) [reply common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java:98]: > 

why unsafe? int type may be enough for some size.

Comment (frankliee) [reply common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java:98]: The max value of int is 4GB, so Linux kernel uses long for all size types (i.e., size_t).

File: common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java [hunk 4]
```diff
@@ -98,6 +98,8 @@ static Long convertToLong(Object o) {
       return ((Integer) o).longValue();
     }
     if (UnitConverter.isByteString(o.toString())) {
+      // Use long type as size type as much as possible,
+      // so UnitConverter only support long type
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java:102]: add the comment `intType only support the size that less than 4GB`

File: common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  public void testByteString() {
+    assertEquals(10 * GB, UnitConverter.byteStringAs("10GB", ByteUnit.BYTE));
+    assertEquals(10 * GB, UnitConverter.byteStringAs("10gb", ByteUnit.BYTE));
+    assertEquals(10 * GB, UnitConverter.byteStringAs("10gB", ByteUnit.BYTE));
+
+    assertEquals(10 * MB, UnitConverter.byteStringAs("10MB", ByteUnit.BYTE));
+    assertEquals(10 * MB, UnitConverter.byteStringAs("10mb", ByteUnit.BYTE));
+    assertEquals(10 * MB, UnitConverter.byteStringAs("10mB", ByteUnit.BYTE));
+
+    assertEquals(10 * KB, UnitConverter.byteStringAs("10KB", ByteUnit.BYTE));
```
Comment (jerqi) [inline common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java:41]: Should we test `TB` and `PB`?

File: common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java (hunk 4 above)
Comment (frankliee) [reply common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java]: fixed

File: common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java (hunk 5 above)
Comment (frankliee) [reply common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java:41]: fixed

File: common/src/main/java/com/tencent/rss/common/rpc/GrpcServer.java [hunk 6]
```diff
@@ -63,14 +63,14 @@ public GrpcServer(RssBaseConf conf, BindableService service, GRPCMetrics grpcMet
...
-          .maxInboundMessageSize(maxInboundMessageSize)
+          .maxInboundMessageSize((int)maxInboundMessageSize)
           .build();
     } else {
       this.server = ServerBuilder
           .forPort(port)
           .addService(service)
           .executor(pool)
-          .maxInboundMessageSize(maxInboundMessageSize)
+          .maxInboundMessageSize((int)maxInboundMessageSize)
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/rpc/GrpcServer.java:73]: we should check `0 < maxInboundMessageSize < Integer.Max_VALUE`

File: common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java (hunk 5 above)
Comment (jerqi) [reply common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java:41]: ??? i seems not be fixed.

Comment (frankliee) [reply common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java:52]: Freshed

File: common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  long KB = (long)ByteUnit.KiB.toBytes(1L);
+
+  @Test
+  public void testByteString() {
+
+    assertEquals(10 * PB, UnitConverter.byteStringAs("10PB", ByteUnit.BYTE));
+    assertEquals(10 * PB, UnitConverter.byteStringAs("10pb", ByteUnit.BYTE));
+    assertEquals(10 * PB, UnitConverter.byteStringAs("10pB", ByteUnit.BYTE));
+
+    assertEquals(10 * TB, UnitConverter.byteStringAs("10TB", ByteUnit.BYTE));
```
Comment (jerqi) [inline common/src/test/java/com/tencent/rss/common/util/UnitConverterTest.java:40]: test `10t`,`10g`,`10T` etc.

File: common/src/main/java/com/tencent/rss/common/rpc/GrpcServer.java (hunk 6 above)
Comment (frankliee) [reply common/src/main/java/com/tencent/rss/common/rpc/GrpcServer.java:73]: fixed by checkValue

File: common/src/main/java/com/tencent/rss/common/config/RssBaseConf.java [hunk 8]
```diff
@@ -114,10 +114,12 @@
       .defaultValue(30 * 1000L)
       .withDescription("jetty http idle timeout (ms) ");
 
-  public static final ConfigOption<Integer> RPC_MESSAGE_MAX_SIZE = ConfigOptions
+  public static final ConfigOption<Long> RPC_MESSAGE_MAX_SIZE = ConfigOptions
       .key("rss.rpc.message.max.size")
-      .intType()
-      .defaultValue(Integer.MAX_VALUE)
+      .longType()
+      .checkValue((v) -> {return v > 0L && v < Integer.MAX_VALUE;},
```
Comment (jerqi) [inline common/src/main/java/com/tencent/rss/common/config/RssBaseConf.java:120]: Extract a common method like `positiveLongValidator`

File: common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java [hunk 9]
```diff
@@ -162,6 +180,9 @@ static Double convertToDouble(Object o) {
 
   public static Function<Long, Boolean> positiveLongValidator = value -> value > 0;
 
+  public static Function<Long, Boolean> positiveIntegerValidator =
+    value -> value > 0L && value < Integer.MAX_VALUE;
```
[the remark this answers is not in the document]
Comment (frankliee) [reply common/src/main/java/com/tencent/rss/common/config/ConfigUtils.java]: done

**Review by jerqi — APPROVED** [review]

**Comment by jerqi** [comment]
Congrats!

**Pull request merged**